Move `forward_htlcs` into standalone lock

ViktorTigerstrom opened the pull request:
Prerequisite for #1507. 

This PR moves the `forward_htlcs` into standalone lock, to make it possible to change the lock order of the `channel_state` lock in #1507.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3874,12 +3872,9 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-	fn fail_htlc_backwards_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_hash: &PaymentHash, onion_error: HTLCFailReason, destination: HTLCDestination) {
+	/// Note that while this function pushes events as well as FailHTLC's to fail htlcs for
+	/// designated channels, no assumptions are made that the channels are still available.
+	fn fail_htlc_backwards_internal(&self, source: HTLCSource, payment_hash: &PaymentHash, onion_error: HTLCFailReason,destination: HTLCDestination) {
```
Wasn't sure if we'd actually like to send in the `forward_htlc` lock here, to ensure that it isn't drained through `process_pending_htlc_forwards` before all htlcs have been processed in `fail_htlc_backwards`, as well as all htlc mpp fails in `claim_funds`. If we'd actually like that, let me know and I'll take the lock before those loops in `fail_htlc_backwards`/`claim_funds` :).

valentinewallace reviewed (commented):
Largely looks good to me. 

@TheBlueMatt I noticed a few undocumented lock order dependencies, such as `pending_outbound_payments` / `pending_events` and `peer_state` / `outbound_scid_aliases`. Not sure how much these matter but could be an opportunity to add docs here

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3095,7 +3095,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			let channel_state = &mut *channel_state_lock;
 
-			for (short_chan_id, mut pending_forwards) in channel_state.forward_htlcs.drain() {
+			for (short_chan_id, mut pending_forwards) in self.forward_htlcs.lock().unwrap().drain() {
```
Good idea! Fixed with 32e3a3c339a5f39bc2fe0f24d680c7824bcce474

ViktorTigerstrom commented:
Thank you very much for the review @valentinewallace & @TheBlueMatt!
I've addressed the feedback with a fixup commit, as well as a separate commit for updating the lock order docs.
I also changed the public accessibility of that `forward_htlcs` field in a fixup commit.

Many of the lock orders are implied through the writer function for the serialization of the ChannelManager currently, but I figured that it's probably worth detailing those lock orders anyways. Let me know if you think differently :). Also let me know if you think the docs commit fits better in a different PR.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -721,9 +716,29 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
+	/// Note that no consistency guarantees are made about the existence of a channel with the
+	/// `short_channel_id` here, nor the `short_channel_id` in the `PendingHTLCInfo`!
+	///
+	/// If also holding `channel_state` lock, must lock `channel_state` prior to `forward_htlcs`.
```
```suggestion
	/// Locked *after* `channel_state`.
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3095,7 +3128,10 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			let channel_state = &mut *channel_state_lock;
 
-			for (short_chan_id, mut pending_forwards) in channel_state.forward_htlcs.drain() {
+			let mut forward_htlcs = HashMap::new();
+			mem::swap(&mut forward_htlcs, &mut self.forward_htlcs.lock().unwrap().drain().collect());
```
```suggestion
			let mut forward_htlcs = HashMap::new();
			mem::swap(&mut forward_htlcs, &mut self.forward_htlcs.lock().unwrap());
```
avoids an allocation

ViktorTigerstrom replied:
Nice, thanks :)!

valentinewallace reviewed (commented):
Thanks for the added docs improvements! 🫡

ViktorTigerstrom commented:
Addressed the latest feedback, and also extended the lock order docs for the `id_to_peer` map.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4140,7 +4173,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					htlc_msat_height_data.extend_from_slice(&byte_utils::be32_to_array(
 							self.best_block.read().unwrap().height()));
-					self.fail_htlc_backwards_internal(channel_state.take().unwrap(),
+					self.fail_htlc_backwards_internal(
```
Thanks! Fixed this with 383ef941a1f99566bab8ffb5e81737b983663a78. Not sure if I overdid it, given that we'll have the unique peer´s  lock with #1507, which kind of removes the purpose of my refactor, instead of just taking the `channel_state` lock several times. I pushed this in a separate commit just in case you think I overdid this, so I can revert it easly.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3751,7 +3789,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				htlc_msat_height_data.extend_from_slice(&byte_utils::be32_to_array(
 						self.best_block.read().unwrap().height()));
-				self.fail_htlc_backwards_internal(channel_state.take().unwrap(),
+				self.fail_htlc_backwards_internal(
```
Fixed with c709bd4951fe447b50416ac52b58f1825f65d8a4

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3095,7 +3130,10 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 
-			for (short_chan_id, mut pending_forwards) in channel_state.forward_htlcs.drain() {
+			let mut forward_htlcs = HashMap::new();
+			mem::swap(&mut forward_htlcs, &mut self.forward_htlcs.lock().unwrap());
```
Fixed with c709bd4951fe447b50416ac52b58f1825f65d8a4. Let me know if you think it's overkill to acquire the lock in the beginning of the `forward_htlcs` loop until #1639 and #1507

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -721,9 +716,29 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
+	/// Note that no consistency guarantees are made about the existence of a channel with the
+	/// `short_channel_id` here, nor the `short_channel_id` in the `PendingHTLCInfo`!
+	///
+	/// Locked *after* `channel_state`.
```
We could, if but that would require that we take the `channel_state` lock temporarily before taking the `forward_htlcs` lock in `ChannelManager::write` function, drop it, and then take it right after again for `ChannelHolder::claimable_htlcs` map (unless we do some `mem::swap`ing). That seems kind of dirty, so I'd suggest we stick to this `lock_order` until  we'd moved `claimable_htlcs` or the channels themselves as well :)?

valentinewallace replied:
Given that's the only place where the lock order arises between `channel_state` and `forward_htlcs`, I'd just throw scopes around the first `channel_state` and `forward_htlcs` usages so we can keep them not dependent on each other in terms of ordering. Don't feel strongly though

ViktorTigerstrom replied:
Fixed with:
4af18a6

ViktorTigerstrom commented:
Thanks for the review @TheBlueMatt! You're right, it makes a lot of sense to do that cleanup here :). Addressed the feedback with the latest commits.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4078,8 +4111,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
 
-		let mut channel_state = Some(self.channel_state.lock().unwrap());
```
Hmmm sorry if I'm misunderstanding this feedback, but I think the current code does handle the claim under one lock or am I missing something? Unless we're not ok with only taking the `channel_state` lock temporarily when removing the `htlc` from `claimable_htlcs`, and taking it later when handling the claim (but that should be ok from my understanding as `claimable_htlcs` have no consistency guarantees), and we're also ok with not holding the `channel_state` lock during `fail_htlc_backwards_internal`.

Please let me know what I'm missing :)!

valentinewallace replied:
I'm curious why it's not okay to take the `channel_state` lock after removing the source too. 

It does kinda seem like the diff could be smaller in `claim_funds` though. If we need to revert this line to take the channel_state again, I think the diff can roughly:

```
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index f52a0414..f2837d5c 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -4156,11 +4156,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			let mut claimed_any_htlcs = false;
 			for htlc in sources.drain(..) {
 				if !valid_mpp {
-					if channel_state.is_none() { channel_state = Some(self.channel_state.lock().unwrap()); }
+					channel_state.take();
 					let mut htlc_msat_height_data = byte_utils::be64_to_array(htlc.value).to_vec();
 					htlc_msat_height_data.extend_from_slice(&byte_utils::be32_to_array(
 							self.best_block.read().unwrap().height()));
-					self.fail_htlc_backwards_internal(channel_state.take().unwrap(),
+					self.fail_htlc_backwards_internal(
 									 HTLCSource::PreviousHopData(htlc.prev_hop), &payment_hash,
 									 HTLCFailReason::Reason { failure_code: 0x4000|15, data: htlc_msat_height_data },
 									 HTLCDestination::FailedPayment { payment_hash } );

ViktorTigerstrom replied:
Thanks for the suggestion @valentinewallace and thanks for the clarification @TheBlueMatt. I kept the patch as is for now, but I'll of course change if you'd prefer :)

ViktorTigerstrom commented:
Addressed the latest feedback, except the "channel_state lock for the entire duration of a claim" feedback above. @TheBlueMatt, please check my comment regarding that and let me know what I'm missing :)!

Also placed moved the fixup commits to be in correct order (under the commit they are supposed to be fixed up into).

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -707,7 +697,12 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
-	/// Locked *after* channel_state.
+	///
+	/// If any of these locks are held at the same time, they must be acquired prior and in the
+	/// following order:
```
Fixed with:
2c5a962

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -721,9 +716,29 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
+	/// `short_channel_id` here, nor the `short_channel_id` in the `PendingHTLCInfo`!
+	///
+	/// Locked *after* `channel_state`.
+	#[cfg(any(test, feature = "_test_utils"))]
```
Fixed with:
425fcc6

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3874,12 +3910,9 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-	fn fail_htlc_backwards_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_hash: &PaymentHash, onion_error: HTLCFailReason, destination: HTLCDestination) {
+	/// Note that while this function pushes events as well as FailHTLC's to fail htlcs for
+	/// designated channels, no assumptions are made that the channels are still available.
+	fn fail_htlc_backwards_internal(&self, source: HTLCSource, payment_hash: &PaymentHash, onion_error: HTLCFailReason,destination: HTLCDestination) {
```
Fixed with:
0ceb5be

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3874,12 +3888,18 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-	/// still-available channels.
-	fn fail_htlc_backwards_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_hash: &PaymentHash, onion_error: HTLCFailReason, destination: HTLCDestination) {
+	/// Note that while this function pushes events as well as FailHTLC's to fail htlcs for
+	/// designated channels, no assumptions are made that the channels are still available.
```
Fixed with:
3c546e71

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -720,10 +723,21 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
 	///
 	/// See `PendingOutboundPayment` documentation for more info.
-	///
-	/// Locked *after* channel_state.
```
Fixed with:
da396f4f

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -781,11 +795,10 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
 	per_peer_state: RwLock<HashMap<PublicKey, Mutex<PeerState>>>,
 
 	pending_events: Mutex<Vec<events::Event>>,
+
```
Fixed with:
da396f4f

ViktorTigerstrom commented:
Addressed the latest feedback with last push, so that you can see the changes easily. Sadly this needs rebase now, so I'll fix the rebase and push that now in a separate push.

ViktorTigerstrom commented:
Rebased on upstream!

ViktorTigerstrom commented:
Snap, noticed that I missed that the `best_block` lock has a lock order after the `per_peer_state` lock. 

The previous lock order comments became really complicated because of that, with all different lock order branches, so I opted to update the docs to instead be illustrated in a tree-like as file directories are often illustrated. I'm open to other suggestions if you think that's a bad idea though :)!

valentinewallace reviewed (commented):
Think I'm ACK after this!

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3972,72 +4014,77 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-						log_error!(self.logger, "Somehow ended up with an MPP spontaneous payment - this should not be reachable!");
+			let mut errs = Vec::new();
+			let mut claimed_any_htlcs = false;
+			{
```
Sure!
[81bf4ac](https://github.com/lightningdevkit/rust-lightning/pull/1656/commits/81bf4ac2b3263af1003e7c93915236a7fb8b92bc)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -677,6 +667,30 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
+// `ChannelManager`. If two or more locks on the same branch is held at the same time, locks should
+// be taken in the order of the lowest to highest level in the tree.
+//
+// Lock order tree:
```
Thanks, awesome to hear :)!

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -677,6 +667,30 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
+// │   │
+// │   └──`per_peer_state`
+// │       │
+// │       │──`best_block`
```
Ah yes you're absolutely correct.. Thanks for the correction, and sorry for the errors with the lock orders! I confusingly thought yesterday that only write locks created lock orders, and not read locks as well.

I've updated the tree with `best_block` after `pending_outbound_payments`, since I've confirmed that it's the case. Also added the `outbound_scid_aliases` lock after the `per_peer_state` in the tree, after confirming that lock order. Finally, I fixed some grammar mistakes + added the `total_consistency_lock` and `forward_htlcs` for the illustration to  be more correct. [aae7b4e](https://github.com/lightningdevkit/rust-lightning/pull/1656/commits/aae7b4eedf3d3a7c39dd023a396dd36133798db3)

In case it's helpful, I've created a simple lock order test for the `ChannelManager` that I think touches most of the lock orders. I'm yet to confirm that it's the case, so it would need more work if we'd want something similar:
```
#[test]
fn lock_order_test() {
let chanmon_cfgs = create_chanmon_cfgs(3);
let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
let nodes = create_network(3, &node_cfgs, &node_chanmgrs);
let chan_1 = create_announced_chan_between_nodes(&nodes, 0, 1, channelmanager::provided_init_features(), channelmanager::provided_init_features());
let chan_2 = create_announced_chan_between_nodes(&nodes, 1, 2, channelmanager::provided_init_features(), channelmanager::provided_init_features());
send_payment(&nodes[0], &vec!(&nodes[1], &nodes[2])[..], 100000);
let _closed_channel_1 = close_channel(&nodes[0], &nodes[1], &chan_1.2, chan_1.3, true);
let _closed_channel_2 = close_channel(&nodes[1], &nodes[2], &chan_2.2, chan_2.3, true);
{
	let _tc_lock = nodes[0].node.total_consistency_lock.write().unwrap();
	let _channel_state_lock = nodes[0].node.channel_state.lock().unwrap();
	let _id_to_peer = nodes[0].node.id_to_peer.lock().unwrap();
	//let _short_to_chan_info = nodes[0].node.short_to_chan_info.write().unwrap();
	let _per_peer_state = nodes[0].node.per_peer_state.write().unwrap();
	let _outbound_scid_aliases = nodes[0].node.outbound_scid_aliases.lock().unwrap();
	let _pending_inbound_payments = nodes[0].node.pending_inbound_payments.lock().unwrap();
	let _pending_outbound_payments = nodes[0].node.pending_outbound_payments.lock().unwrap();
	let _best_block = nodes[0].node.best_block.write().unwrap();
	let _pending_events = nodes[0].node.pending_events.lock().unwrap();
	let _pending_background_events = nodes[0].node.pending_background_events.lock().unwrap();
	let _forward_htlcs = nodes[0].node.forward_htlcs.lock().unwrap();
}
nodes[0].node.get_and_clear_pending_events();
nodes[1].node.get_and_clear_pending_events();
nodes[2].node.get_and_clear_pending_events();
} 
```

valentinewallace reviewed (commented):
LGTM after squash and a second reviewer

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -677,6 +667,36 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
+//
+// Lock order:
+// The tree structure below illustrates the lock order requirements for the different locks of the
+// `ChannelManager`. If two or more locks on the same branch are held at the same time, locks
```
I would prefer compile-time checks too, couldn't find an obvious way though :/ Asked on the rust discord: https://discord.com/channels/442252698964721669/448238009733742612/1020827841564266606 which didn't really bear fruit and separately found this project: https://github.com/BurtonQin/lockbud which may be worth looking into 🤷‍♀️

ViktorTigerstrom replied:
Oh yeah, agree that the docs we're very unclear in regards to taking locks across branches at the same time. I've attempted to make them more clear with: 
0d43f16
Let me know if you still find them unclear.

Also, interesting find @valentinewallace. If you both think it's worth the time researching more about compile-time checks, It's of course something I could do :).

valentinewallace replied:
Ah no worries @ViktorTigerstrom, I forgot we had built the `debug_sync::Mutex` which is literally designed for lock order inversion detection like Matt said

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -677,6 +667,36 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
+// Lock order tree:
+//
+// `total_consistency_lock`
+//  │
```
Replaced the unicode with ASCII with [d1e43f3](https://github.com/lightningdevkit/rust-lightning/pull/1656/commits/d1e43f392ca8ae45a09c9f0ffd818046f80097b6) .

Also removed the extra lines with: [5d5a813](https://github.com/lightningdevkit/rust-lightning/pull/1656/commits/5d5a813ed2ff76e695d81e4819a2ee8c931562b4)
I did that one in a specific fixup just so you can compare the difference to see if you think it affects the readability of the tree. I personally think removing the the extra lines makes the tree considerably less readable, but can also see the point in removing them. Let me know which version you prefer!

ViktorTigerstrom commented:
Squashed and addressed the new feedback in new fixup commits.

valentinewallace reviewed (dismissed):
LGTM after squash

ViktorTigerstrom commented:
Squashed! I dropped the commit that removed the empty rows now as I think the tree becomes less readable without them. But I'll of course remove them again if @valentinewallace also prefers the tree without the empty rows.

valentinewallace reviewed (approved):
LGTM 🌟 

No strong feelings on the spacing, I guess I slightly prefer more concise too but fine as is
